File: Python/2181.py
# 2181. Merge Nodes in Between Zeros
# https://leetcode.com/problems/merge-nodes-in-between-zeros/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
  def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
    back = head
    front = head.next

    while front is not None:
      if front.next is None:
        back.next = None
        break
      elif front.val == 0:
        back.next = front
        back = back.next
      else:
        back.val += front.val

      front = front.next

    return head

File: Rust/2181.rs
/**
 * 2181. Merge Nodes in Between Zeros
 * https://leetcode.com/problems/merge-nodes-in-between-zeros/
 */

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

impl Solution {
  pub fn merge_nodes(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    let mut front = head.as_ref();
    let mut head = Box::new(ListNode::new(0));
    let mut back = &mut head;

    while let Some(node) = front {
      front = node.next.as_ref();

      if node.next.is_none() {
        break;
      } else if node.val == 0 {
        back.next = Some(Box::new(ListNode::new(0)));
        back = back.next.as_mut().unwrap();
      } else {
        back.val += node.val;
      }
    }

    return head.next;
  }
}
